Sources/ClientRuntime/Message/URI.swift

```diff
@@ -0,0 +1,180 @@
+        self.urlComponents.host = value
+        return self
+    }
+
+    @discardableResult
+    public func withPort(_ value: Int16?) -> URIBuilder {
+        self.urlComponents.port = value.map { Int($0) }
+        return self
+    }
+
+    @discardableResult
+    public func withPort(_ value: Int?) -> URIBuilder {
+        self.urlComponents.port = value
+        return self
+    }
+
+    @discardableResult
+    public func withQueryItems(_ value: [SDKURLQueryItem]) -> URIBuilder {
+        self.urlComponents.percentEncodedQueryItems = value.isEmpty ? nil : value.toURLQueryItems
+        return self
+    }
+
+    @discardableResult
+    public func appendQueryItems(_ items: [SDKURLQueryItem]) -> URIBuilder {
+        guard !items.isEmpty else {
+            return self
+        }
+        var queryItems = self.urlComponents.percentEncodedQueryItems ?? []
+        queryItems += items.toURLQueryItems
+        self.urlComponents.percentEncodedQueryItems = queryItems
+        return self
+    }
+
+    @discardableResult
+    public func appendQueryItem(_ item: SDKURLQueryItem) -> URIBuilder {
+        self.appendQueryItems([item])
+        return self
+    }
+
+    @discardableResult
+    public func withUsername(_ value: String?) -> URIBuilder {
+        self.urlComponents.user = value
+        return self
+    }
+
+    @discardableResult
+    public func withPassword(_ value: String?) -> URIBuilder {
+        self.urlComponents.password = value
+        return self
+    }
+
+    public func build() -> URI {
+        return URI(scheme: Scheme(rawValue: self.urlComponents.scheme!)!,
+                   path: self.urlComponents.percentEncodedPath,
+                   host: self.urlComponents.host!,
+                   port: self.urlComponents.port.map { Int16($0) },
+                   queryItems: self.urlComponents.percentEncodedQueryItems?.map {
+                        SDKURLQueryItem(name: $0.name, value: $0.value)
+                   } ?? [],
+                   username: self.urlComponents.user,
+                   password: self.urlComponents.password)
+    }
+
+    // We still have to keep 'url' as an optional, since we're
+    // dealing with dynamic components that could be invalid.
+    fileprivate func getUrl() -> URL? {
+        let isInvalidHost = self.urlComponents.host?.isEmpty ?? false
+        return isInvalidHost && self.urlComponents.path.isEmpty ? nil : self.urlComponents.url
+    }
+}
+
+extension String {
+    var isPercentEncoded: Bool {
+        let decoded = self.removingPercentEncoding
+        return decoded != nil && decoded != self
+    }
+}
+
+extension Array where Element == SDKURLQueryItem {
+    public var queryString: String? {
+        if self.isEmpty {
+            return nil
+        }
+        return self.map { [$0.name, $0.value].compactMap { $0 }.joined(separator: "=") }.joined(separator: "&")
+    }
+
+    public var toURLQueryItems: [URLQueryItem] {
```

Comment:
Nit: naming of `to` sounds like a method, maybe this makes more sense as a method anyway, not sure. Same thing for `toSDKURLQueryItems` below

Reply:
Updated to functions